lib/Dialect/ESI/ESIPasses.cpp

```diff
@@ -0,0 +1,106 @@
+//===- ESIToRTL.cpp - ESI to RTL/SV conversion passes -----------*- C++ -*-===//
+//
+// Lower ESI to RTL and SV.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/ESI/ESIDialect.h"
+#include "circt/Dialect/ESI/ESIOps.h"
+#include "circt/Dialect/RTL/Dialect.h"
+#include "circt/Dialect/SV/Dialect.h"
+
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/Twine.h"
+#include "llvm/Support/Debug.h"
+
+#include <memory>
+
+namespace circt {
+namespace esi {
+#define GEN_PASS_CLASSES
+#include "circt/Dialect/ESI/ESIPasses.h.inc"
+} // namespace esi
+} // namespace circt
+
+using namespace mlir;
+using namespace circt::esi;
+
+namespace {
+
+/// Lower `ChannelBuffer`s, breaking out the various options. For now, just
+/// replace with the specified number of pipeline stages (since that's the only
+/// option).
+struct ChannelBufferLowering : public ConversionPattern {
+  ChannelBufferLowering(MLIRContext *ctx)
+      : ConversionPattern(ChannelBuffer::getOperationName(), 1, ctx) {}
+
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const final {
+    auto loc = op->getLoc();
+    ChannelBuffer buffer;
+    if (!(buffer = dyn_cast<ChannelBuffer>(op)))
```

Comment:
nit, but I'd write this as:

```
auto buffer = dyn_cast<ChannelBuffer>(op);
if (!buffer) return failure();
```

Reply:
nit on a nit but I find the way this is written now harder to read than Chris' comment. I didn't look at the original.